src/bootchooser.c

```diff
@@ -975,6 +978,435 @@ static gboolean uboot_set_primary(RaucSlot *slot, GError **error)
+	if (!raspberrypi_load_autoboot(r_context()->config->autoboottxt_path, &key_file, &ierror)) {
+		g_propagate_error(error, ierror);
+		return NULL;
+	}
+
+	boot_partition = g_key_file_get_integer(key_file, "all", "boot_partition", &ierror);
+	if (ierror) {
+		g_set_error(
+				error,
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_PARSE_FAILED,
+				"Property boot_partition not set in %s", filename);
+		return NULL;
+	}
+
+	return raspberrypi_find_config_slot_by_boot_partition(r_context()->config, boot_partition);
+}
+
+/* Set the oneshot reboot flag to cause the firmware to run tryboot at next
+ * reboot.
+ *
+ * The firmware uses the boot_partition defined in the [tryboot] section and it
+ * loads the alternate configuration file tryboot.txt instead of config.txt at
+ * next boot. */
+static gboolean raspberrypi_set_primary_temporary(GError **error)
+{
+	GError *ierror = NULL;
+
+	if (!raspberrypi_tryboot_set(TRUE, error)) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to set reboot flag: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* Swap the boot_partition configurations in autoboot.txt to cause the firmware
+ * to use the boot_partition that is defined in the [tryboot] section as the
+ * default boot_partition. */
+static gboolean raspberrypi_set_primary_persistent(GError **error)
+{
+	g_autoptr(GKeyFile) key_file = NULL;
+	GError *ierror = NULL;
+	const gchar *filename;
+	gint boot_partition;
+	gint tryboot_boot_partition;
+
+	filename = r_context()->config->autoboottxt_path;
+	if (!raspberrypi_load_autoboot(r_context()->config->autoboottxt_path, &key_file, &ierror)) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	boot_partition = g_key_file_get_integer(key_file, "all", "boot_partition", &ierror);
+	if (ierror) {
+		g_set_error(
+				error,
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_PARSE_FAILED,
+				"Property boot_partition not set in section [all] in file %s", filename);
+		return FALSE;
+	}
+
+	tryboot_boot_partition = g_key_file_get_integer(key_file, "tryboot", "boot_partition", &ierror);
+	if (ierror) {
+		g_set_error(
+				error,
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_PARSE_FAILED,
+				"Property boot_partition not set in section [tryboot] in file %s", filename);
+		return FALSE;
+	}
+
+	if (!g_key_file_remove_key(key_file, "all", "boot_partition", &ierror)) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to remove boot_partition key in section [all]: ");
+		return FALSE;
+	}
+
+	if (!g_key_file_remove_key(key_file, "tryboot", "boot_partition", &ierror)) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to remove boot_partition key: ");
+		return FALSE;
+	}
+
+	g_key_file_set_integer(key_file, "all", "boot_partition", tryboot_boot_partition);
+	g_key_file_set_integer(key_file, "tryboot", "boot_partition", boot_partition);
+	if (!raspberrypi_save_autoboot_temporary(key_file, &ierror)) {
```

Comment:
Calling `raspberrypi_save_autoboot_temporary` from `raspberrypi_set_primary_persistent` is confusing, please rename.

Reply:
Okay. I have merged the function `raspberrypi_save_autoboot_temporary()` and `raspberrypi_save_autoboot()` in commit e77fce6826f25ea6c70ec717c09034c0c5706109.

There is no more `_temporary`, it should be less confusing now.